src/cpu/pred/bpred_unit.cc

```diff
@@ -419,105 +534,206 @@ BPredUnit::squash(const InstSeqNum &squashed_sn,
+
+            DPRINTF(Branch,"[tid:%i] [squash sn:%llu] "
+                "BTB miss PC %#x %s \n", tid, squashed_sn,
+                hist->pc, toString(hist->type));
         }
+
     } else {
         DPRINTF(Branch, "[tid:%i] [sn:%llu] pred_hist empty, can't "
                 "update\n", tid, squashed_sn);
     }
 }
 
+
 void
 BPredUnit::dump()
 {
     int i = 0;
     for (const auto& ph : predHist) {
         if (!ph.empty()) {
-            auto pred_hist_it = ph.begin();
+            auto hist = ph.begin();
 
             cprintf("predHist[%i].size(): %i\n", i++, ph.size());
 
-            while (pred_hist_it != ph.end()) {
+            while (hist != ph.end()) {
                 cprintf("sn:%llu], PC:%#x, tid:%i, predTaken:%i, "
-                        "bpHistory:%#x\n",
-                        pred_hist_it->seqNum, pred_hist_it->pc,
-                        pred_hist_it->tid, pred_hist_it->predTaken,
-                        pred_hist_it->bpHistory);
-                pred_hist_it++;
+                        "bpHistory:%#x, rasHistory:%#x\n",
+                        (*hist)->seqNum, (*hist)->pc,
+                        (*hist)->tid, (*hist)->predTaken,
+                        (*hist)->bpHistory, (*hist)->rasHistory);
+                hist++;
             }
 
             cprintf("\n");
         }
     }
 }
 
+
+BPredUnit::BPredUnitStats::BPredUnitStats(
+                                statistics::Group *parent, BPredUnit *bp)
+    : statistics::Group(parent),
+      ADD_STAT(lookups, statistics::units::Count::get(),
+              "Number of BP lookups"),
+      ADD_STAT(squashes, statistics::units::Count::get(),
+              "Number of branches that got squashed as an earlier branch was "
+              "mispredicted."),
+      ADD_STAT(corrected, statistics::units::Count::get(),
+              "Number of branches that got corrected but not yet commited. "),
+      ADD_STAT(committed, statistics::units::Count::get(),
+              "Number of branches finally committed "),
+      ADD_STAT(mispredicted, statistics::units::Count::get(),
+              "Number of committed branches that where mispredicted."),
+      ADD_STAT(targetProvider, statistics::units::Count::get(),
+              "Number of branches commited per branch type"),
+      ADD_STAT(targetWrong, statistics::units::Count::get(),
+              "Number of branches commited per branch type"),
+      ADD_STAT(earlyResteers, statistics::units::Count::get(),
+              "Number of branches that got squashed after decode."),
+
+      ADD_STAT(condPredicted, statistics::units::Count::get(),
+               "Number of conditional branches predicted"),
+      ADD_STAT(condPredictedTaken, statistics::units::Count::get(),
+               "Number of conditional branches predicted as taken"),
+      ADD_STAT(condIncorrect, statistics::units::Count::get(),
+               "Number of conditional branches incorrect"),
+      ADD_STAT(incorrect, statistics::units::Count::get(),
```

Comment:
TODO: cleanup stats.
incorrect is the same as condIncorrect.
Also we should maybe remove some stats.